spark2/src/test/java/org/apache/iceberg/spark/source/TestStructuredStreamingRead.java

```diff
@@ -0,0 +1,522 @@
+        IllegalStateException.class, "The option starting-snapshot-id -1 is not an ancestor",
+        () -> source.createMicroBatchReader(Optional.empty(), checkpoint.toString(), options));
+  }
+
+  @SuppressWarnings("unchecked")
+  @Test
+  public void testSpecifySnapshotId() throws IOException {
+    File checkpoint = Files.createTempDirectory(parent, "checkpoint").toFile();
+
+    IcebergSource source = new IcebergSource();
+    List<Long> snapshotIds = SnapshotUtil.currentAncestors(table);
+    Collections.reverse(snapshotIds);
+
+    // test specify snapshot-id
+    DataSourceOptions options = new DataSourceOptions(ImmutableMap.of(
+        "path", tableLocation.toString(),
+        "checkpointLocation", checkpoint.toString(),
+        "starting-snapshot-id", snapshotIds.get(1).toString(),
+        "max-size-per-batch", "1000"));
+    StreamingReader streamingReader = (StreamingReader) source.createMicroBatchReader(
+        Optional.empty(), checkpoint.toString(), options);
+
+    streamingReader.setOffsetRange(Optional.empty(), Optional.empty());
+    StreamingOffset start = (StreamingOffset) streamingReader.getStartOffset();
+    validateOffset(new StreamingOffset(snapshotIds.get(1), 0, true, false), start);
+
+    StreamingOffset end = (StreamingOffset) streamingReader.getEndOffset();
+    validateOffset(new StreamingOffset(snapshotIds.get(1), 1, true, false), end);
+
+    streamingReader.setOffsetRange(Optional.of(end), Optional.empty());
+    StreamingOffset end1 = (StreamingOffset) streamingReader.getEndOffset();
+    validateOffset(new StreamingOffset(snapshotIds.get(2), 1, false, true), end1);
+
+    streamingReader.setOffsetRange(Optional.of(end1), Optional.empty());
+    StreamingOffset end2 = (StreamingOffset) streamingReader.getEndOffset();
+    validateOffset(new StreamingOffset(snapshotIds.get(3), 1, false, true), end2);
+
+    streamingReader.setOffsetRange(Optional.of(end2), Optional.empty());
+    StreamingOffset end3 = (StreamingOffset) streamingReader.getEndOffset();
+    validateOffset(end2, end3);
+  }
+
+  @Test
+  public void testStreamingRead() {
+    Dataset<Row> read = spark.readStream()
+        .format("iceberg")
+        .load(tableLocation.toString());
+    DataStreamWriter<Row> streamWriter = read.writeStream()
+        .format("memory")
+        .outputMode("append")
+        .queryName("memoryStream");
+
+    try {
+      StreamingQuery query = streamWriter.start();
+      query.processAllAvailable();
+
+      List<SimpleRecord> actual = spark.table("memoryStream")
+          .as(Encoders.bean(SimpleRecord.class))
+          .collectAsList();
+      List<SimpleRecord> expectedResult = expected.stream().flatMap(List::stream).collect(Collectors.toList());
+      validateResult(expectedResult, actual);
+    } finally {
+      for (StreamingQuery query : spark.streams().active()) {
+        query.stop();
+      }
+    }
+  }
+
+  @Test
+  public void testStreamingReadWithSpecifiedSnapshotId() {
```

Comment:
There are several other options that can be passed through and we probably need to verify that all of them work, I counted

SPLIT_SIZE
SPLIT_OPEN_FILE_COST
SPLIT_LOOKBACK

Reply:
All these parameters are verified by other tests, and has been tested in Spark batch read test, there's no need to test again.

Reply:
I would feel more comfortable if there were tests here as well since we may not always have the same implementation under the hood. It's also safest to make sure that the options can be properly propagated all the way through, just incase the underlying implementation works, but the passing the parameters through is bugged in some way.